File: src/api/schemas/portfolio.py
"""19번 — 포트폴리오 API 요청 스키마."""
from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel

from src.services.portfolio_service import RebalanceAdjustment


class RebalanceAdjustmentRequest(BaseModel):
    execution_id: int
    new_allocated_capital: Decimal


class RebalanceRequest(BaseModel):
    adjustments: list[RebalanceAdjustmentRequest]


def to_adjustments(body: RebalanceRequest) -> list[RebalanceAdjustment]:
    return [
        RebalanceAdjustment(
            execution_id=a.execution_id, new_allocated_capital=a.new_allocated_capital
        )
        for a in body.adjustments
    ]

File: packages/shared-types/src/portfolio.ts
// src/services/portfolio_service.py 1:1 대응.

export interface PortfolioAllocation {
  executionId: number;
  strategyId: string;
  strategyVersion: string;
  exchange: string;
  mode: string;
  status: string;
  allocatedCapital: string;
  totalPnl: string;
  currentValue: string;
  weightPct: string;
}

export interface PortfolioView {
  allocations: PortfolioAllocation[];
  unallocatedCash: string;
  unallocatedCashWeightPct: string;
  totalPortfolioValue: string;
}

export interface RebalanceAdjustmentRequest {
  executionId: number;
  newAllocatedCapital: string;
}

export interface RebalanceRequest {
  adjustments: RebalanceAdjustmentRequest[];
}

export interface RebalanceResult {
  adjusted: number;
  pendingApproval: number;
  approvalRequestIds: number[];
}

export interface DailyPnL {
  tradeDate: string;
  dailyPnl: string;
  cumulativePnl: string;
}

export interface StrategyContribution {
  strategyId: string;
  strategyVersion: string;
  realizedPnl: string;
  tradeCount: number;
}

export interface ReportSummary {
  periodStart: string;
  periodEnd: string;
  totalReturn: string;
  winRate: string | null;
  maxDrawdown: string;
  tradeCount: number;
  strategyContributions: StrategyContribution[];
  dailyPnl: DailyPnL[];
}
